docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+
+“Reverse” interoperability (using Swift APIs from C++) is another extremely important part of the interoperability story, however, reverse interoperability has largely different goals and constraints, which necessarily mean a different design and roadmap. Alex Lorenz will be putting together a roadmap for reverse interoperability which outlines these goals and will shed light onto the differences between the two halves of the projects. In the future this roadmap will need to be updated to explain how these two halves fit together and present the “bi-directional” interoperability story that is C++ and Swift interoperability as a whole. 
+
+**C++, as a language:** C++ is an stubbornly un-opinionated language. There is little to no consensus on anything from naming, to generic programming, value categories, or error handling (to name a few things). This lack of opinion is in sharp contrast to Swift where there is generally one, specific way to do things. Even in the C++ Standard Library there is a rather un-standard set of API models. The un-opinionated nature of C++ means there are few, if any guarantees about a given C++ API, because the same functionality may be implemented using many different language features or programming paradigms. When consuming an arbitrary C++ API, you must be prepared for *anything*. When consuming Swift APIs, this is not the case. Swift is an opinionated language which means rather than consuming *any* API pattern you are consuming *some*, specific API pattern. Swift clearly defines patterns for naming, generic programming, value categories, error handling, etc. and it is rare that a codebase deviates from the standard practice. 
+
+**C and Objective-C interoperability:** C++ interoperability will have to build off of the previous art of C and Objective-C interoperability. Starting with the name, C++ is a superset of C, which means C++ interoperability has a great starting place to build off of. C is a simple language without many high level constructs, so importing C APIs is, relatively, simple. On the other hand, Objective-C adds some high level constructs, such as reference types and a generics model. However, unlike C++, Objective-C has strong idioms; there are only so many patterns for doing some specific thing. Further, Objective-C was able to be updated to better interoperate with Swift, the generics model is a good example of this. So, while C and Objective-C interoperability is a good starting place, the model for importing these languages cannot be taken wholesale for C++ interoperability. 
+
+## Goals
+
+Safety is a top priority. The compiler should not import APIs that would be less safe in Swift than they are in C++ and interop should strive to make imported APIs *safer* in Swift than they are in C++ by providing safe API interfaces for common, unsafe C++ API patterns (such as iterators).
```

Comment:
Since this is one of the top goals, I think it might be better to explain what "less safe" means in some more detail. E.g. here is a thought  (not sure if fully captures it)
"A C++ API is less safe in Swift, if a Swift code snippet that can be written using the C++ API is invalid in Swift (e.g. due to memory corruption bugs) but its equivalent C++ code snippet is valid in C++ <link to examples that come later in the document>".

Reply:
Not sure if we should talk about "Correctness". It is an implicit goal i.e., equivalent Swift and C++ code snippets have equivalent I/O behavior. But, in a way we rely on validity of user annotations and reasonable expectations on copy constructors, such as copy-elision optimizations on compilers do.

Reply:
Thinking this through, consider:

```cxx
void f(void(*)(int&, int&));
```

If this is imported into Swift as

```swift
f(thin (inout Int, inout Int)->Void)
```

it has become less-safe, because C++ could create a law-of-exclusivity violation.  I guess that means we'd need to import it as

```swift
f(thin (UnsafeMutablePointer<Int>, UnsafeMutablePointer<Int>)->Void)
```

Reply:
I think it's worth creating a C++ sanitizer that could be used to make crossing the boundary between C++ and Swift safer, by dynamically upholding the guarantees that Swift code statically expects (like exclusivity of `inout` access).

Reply:
> The compiler should not import APIs that would be less safe in Swift than they are in C++ and interop should strive to make imported APIs *safer* in Swift than they are in C++ by providing safe API interfaces for common, unsafe C++ API patterns (such as iterators). 

Based on all the discussions, it is already looking like this goal is "best effort", right? The current proposal disallows projections for a very specific definition of projection (using "pointer types"), and it is possible to write valid C++ code that is allowed by the model but more unsafe in Swift than in C++.  So it may be better to rephrase this goal as best effort e.g. may be "the compile should aim to not increase the unsafely when C++ APIs are used from Swift", because completely eliminating it seems like a very hard goal to achieve.